tests/core/utils/urlutils.js

```diff
@@ -0,0 +1,111 @@
+import SearchParams from '../../../src/ui/dom/searchparams';
+import { PRODUCTION, SANDBOX } from '../../../src/core/constants';
+import {
+  getLiveApiUrl,
+  getCachedLiveApiUrl,
+  getKnowledgeApiUrl,
+  getAnalyticsUrl,
+  addParamsToUrl,
+  urlWithoutQueryParamsAndHash,
+  equivalentParams
+} from '../../../src/core/utils/urlutils';
+
+describe('getUrlFunctions work', () => {
+  it('differentiates sandbox from prod', () => {
+    expect(getLiveApiUrl()).not.toEqual(expect.stringContaining('sandbox'));
+    expect(getCachedLiveApiUrl()).not.toEqual(expect.stringContaining('sandbox'));
+    expect(getKnowledgeApiUrl()).not.toEqual(expect.stringContaining('sandbox'));
+    expect(getAnalyticsUrl()).not.toEqual(expect.stringContaining('sandbox'));
+
+    expect(getLiveApiUrl(SANDBOX)).toEqual(expect.stringContaining('sandbox'));
+    expect(getCachedLiveApiUrl(SANDBOX)).toEqual(expect.stringContaining('sandbox'));
+    expect(getKnowledgeApiUrl(SANDBOX)).toEqual(expect.stringContaining('sandbox'));
+    expect(getAnalyticsUrl(SANDBOX)).toEqual(expect.stringContaining('sandbox'));
+  });
+
+  it('differentiates conversion tracking in analytics url', () => {
+    expect(getAnalyticsUrl(PRODUCTION, true)).toEqual(expect.stringContaining('realtimeanalytics'));
+    expect(getAnalyticsUrl(SANDBOX, true)).toEqual(expect.stringContaining('realtimeanalytics'));
+
+    expect(getAnalyticsUrl(PRODUCTION)).not.toEqual(expect.stringContaining('realtimeanalytics'));
+    expect(getAnalyticsUrl(SANDBOX)).not.toEqual(expect.stringContaining('realtimeanalytics'));
+  });
+});
+
+describe('addParamsToUrl works', () => {
+  it('adds params when currentParams is undefined', () => {
+    expect(addParamsToUrl('https://yext.com/', { referrerPageUrl: '' }))
+      .toEqual('https://yext.com/?referrerPageUrl=');
+  });
+
+  it('adds params when params already exist', () => {
+    expect(addParamsToUrl(
+      'https://yext.com/',
+      { query: 'all', referrerPageUrl: '' },
+      new SearchParams('?page=10&facets=true')
```

Comment:
nit: Since SearchParams takes "window.location.search.substring(1)" in the urlutils.js code, should this have the question mark prefix? Functionally it may not make a difference?

Reply:
These should be functionally the same! URLSearchParams will strip any leading `?`.